File: ios/HybridAudioWaveform.swift
import Foundation
import NitroModules

class HybridAudioWaveform: HybridAudioWaveformSpec {
  private let extractor = WaveformExtractor()

  func extractWaveform(path: String, sampleCount: Double) throws -> Promise<[Double]> {
    return Promise.async { [extractor] in
      let samples = try extractor.extract(path: path, sampleCount: Int(sampleCount))
      return samples.map(Double.init)
    }
  }
}

File: android/src/main/java/com/margelo/nitro/audiowaves/HybridAudioWaveform.kt
package com.margelo.nitro.audiowaves

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

class HybridAudioWaveform : HybridAudioWaveformSpec() {
  private val extractor by lazy {
    val context = NitroModules.applicationContext
      ?: throw AudioWavesError.DecodeFailed("NitroModules.applicationContext is null")
    WaveformExtractor(context)
  }

  override fun extractWaveform(path: String, sampleCount: Double): Promise<DoubleArray> {
    return Promise.async {
      extractor.extract(path, sampleCount.toInt())
        .map { it.toDouble() }
        .toDoubleArray()
    }
  }
}